### PR TITLE
Laboratory work #2, Anastasiia Seliverstova - 23FPL1 

### PR DESCRIPTION
**Pull request by seliverstovaa**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,6 +19,23 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+    tokens = []
+    word = []
+    only_letters = []
+    for elem in text.lower():
+        if elem.isalpha():
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:28]: use continue if this condition is not True

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -33,6 +51,21 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(tokens, list) or not isinstance(stopwords, list):
+        return None
+    if len(tokens) == 0 or len(stopwords) == 0:
+        return None
+    for elem in stopwords:
+        if not isinstance(elem, str):
+            return None
+    tokens_without_stopwords = []
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:61]: try to use list comprehensions

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -62,6 +106,18 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
...
+        return None
+    for word in vocab:
+        if not isinstance(word, str):
+            return None
+    for word in document_tokens:
+        if not isinstance(word, str):
+            return None
+        if word not in vocab:
+            vocab.append(word)
+    return {word: document_tokens.count(word)/len(document_tokens) for word in set(vocab)}
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:120]: great!

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -77,6 +133,29 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+    for word in vocab:
+        if not isinstance(word, str):
+            return None
+    idf = {}
+    tokens = []
+    for doc in documents:
+        if not isinstance(doc, list):
+            return None
+        for word in doc:
+            freq = doc.count(word)
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:148]: don't create separate variable

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -120,6 +212,31 @@ def calculate_bm25(
...
+            not isinstance(vocab[0], str) or not isinstance(document[0], str)):
+        return None
+    if (not isinstance(idf_document, dict) or not idf_document
+            or not isinstance(list(idf_document.values())[0], float)):
+        return None
+    if not isinstance(k1, float) or not isinstance(b, float):
+        return None
+    bm25_dict = {}
+    for word in document:
+        if word not in vocab:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:229]: use continue

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great job! Raise your score to 10! I wish you good luck!

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py:229]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py:148]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py:61]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py:28]: After "if not elem.isalpha()" there should be the next step, which is now after else. How can I use continue here?

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -18,6 +19,23 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+    tokens = []
+    word = []
+    only_letters = []
```
[the remark this answers is not in the document]
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py:26]: done

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -33,6 +49,18 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+        return None
+    if len(tokens) == 0 or len(stopwords) == 0:
+        return None
+    for elem in stopwords:
+        if not isinstance(elem, str):
+            return None
+    for elem in tokens:
+        if not isinstance(elem, str):
+            return None
+    tokens_without_stopwords = [elem for elem in tokens if elem not in stopwords]
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:62]: great!

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -47,6 +75,17 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+    if not isinstance(documents, list) or len(documents) == 0:
+        return None
+    all_tokens = []
+    for tokens in documents:
+        if not isinstance(tokens, list):
+            return None
+        for elem in tokens:
+            if not isinstance(elem, str):
+                return None
+        all_tokens.extend(tokens)
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:87]: try to simplify and to use list comprehensions

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -92,6 +165,19 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
...
+        return None
+    if len(tf) == 0 or len(idf) == 0:
+        return None
+    tf_idf = {}
+    for word, value in tf.items():
+        value_idf = idf.get(word)
+        if (not isinstance(word, str) or not isinstance(value, float)
+                or not isinstance(value_idf, float)):
+            return None
+        tf_idf_word = value * value_idf
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:178]: don't create a separate variable

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -138,6 +250,28 @@ def rank_documents(
...
+    result = []
+    for index, story in enumerate(indexes):
+        value = 0.0
+        for word in correct_query:
+            if word not in story.keys():
+                continue
+            num = story.get(word)
+            if not isinstance(num, float):
+                return None
+            value += num
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:271]: can we use sum() here? think about it

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great progress! Now all that's left is to fix the comments!

**Comment by Vasilisa-Blyudova** [comment]
All you have to do is fix the comments! Good luck!

File: lab_2_retrieval_w_bm25/main.py (hunk 9 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 10 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 8 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -77,6 +122,23 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+            or not documents):
+        return None
+    if not all(isinstance(word, str) for doc in documents for word in doc):
+        return None
+    idf = {}
+    for word in vocab:
+        num = 0
+        for doc in documents:
+            if word in doc:
+                num += 1
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:138]: how else can we count the number of elements included in doc?

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great work! All you have to do is fix the comments and tests

**Comment by Vasilisa-Blyudova** [comment]
Unfortunately, I don't see any progress in your PR since the last review.

File: lab_2_retrieval_w_bm25/main.py [hunk 12]
```diff
@@ -18,6 +20,20 @@ def tokenize(text: str) -> list[str] | None:
...
+        tokens.append(elem)
+        if elem.isalpha():
+            continue
+        tokens[tokens.index(elem)] = " "
+    tokens = ''.join(tokens).split(' ')
+    while '' in tokens:
+        tokens.remove('')
+    if not isinstance(tokens, list):
+        return None
+    return tokens
```
[the remark this answers is not in the document]
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py:36]: fixed

File: lab_2_retrieval_w_bm25/main.py (hunk 11 above)
Comment (seliverstovaa) [reply lab_2_retrieval_w_bm25/main.py:136]: I have no idea((

File: lab_2_retrieval_w_bm25/main.py [hunk 13]
```diff
@@ -22,11 +22,16 @@ def tokenize(text: str) -> list[str] | None:
...
-            text = text.replace(char, ' ')
-    return text.lower().split()
+    tokens = []
+    text = text.lower()
+    for elem in text:
+        if not elem.isalpha():
+            elem = elem.replace(elem, ' ')
+        tokens.append(elem)
+    tokens = ''.join(tokens).split(' ')
+    while '' in tokens:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:32]: It is a bad idea iterate over sequence that you are modifing inside the loop

File: lab_2_retrieval_w_bm25/start.py [hunk 14]
```diff
@@ -31,121 +32,67 @@ def main() -> None:
...
+                                        [golden[0] for golden in rank_from_json])
+    spearman_bm_25_adv = calculate_spearman([num[0] for num in rank_from_json],
+                                            [golden[0] for golden in rank_from_json])
+    print(f'Golden rank: {[golden[0] for golden in rank_from_json]}')
+    print(f'Spearman tf_idf: {spearman_tf_idf}')
+    print(f'Spearman BM25: {spearman_bm_25}')
+    print(f'Spearman BM25 with cutoff: {spearman_bm_25_adv}')
+    result = spearman_bm_25_adv
     assert result, "Result is None"
+    return None
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/start.py:95]: no need this return, actually

File: lab_2_retrieval_w_bm25/start.py [hunk 15]
```diff
@@ -31,121 +32,67 @@ def main() -> None:
...
+    rank_tf_idf = rank_documents(metrics_tf_idf,
+                                 'Which fairy tale has Fairy Queen?',
+                                 stopwords)
+    rank_bm_25 = rank_documents(metrics_bm_25,
+                                'Which fairy tale has Fairy Queen?',
+                                stopwords)
+    save_index(bm_25_advanced, "assets/metrics.json")
+    loaded = load_index("assets/metrics.json")
+    if not isinstance(loaded, list):
+        return None
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/start.py:79]: just return, without None

File: lab_2_retrieval_w_bm25/start.py [hunk 16]
```diff
@@ -1,7 +1,8 @@
 """
 Laboratory Work #2 starter
 """
-# pylint:disable=too-many-locals, unused-argument, unused-variable, too-many-branches, too-many-statements, duplicate-code
+# pylint:disable=too-many-locals, unused-argument, unused-variable,
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/start.py:4]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**